File: ruby/find-mode-in-binary-search-tree.rb
# frozen_string_literal: true

# https://leetcode.com/problems/find-mode-in-binary-search-tree/

require_relative "lib/tree_node"

def dfs(node, &)
  Enumerator.new do |traverse|
    recursive_dfs(node) { traverse.yield(_1) }
  end.each(&)
end

def recursive_dfs(node, &block)
  return if node.nil?

  recursive_dfs(node.left, &block)
  block.call(node)
  recursive_dfs(node.right, &block)
end

# @param {TreeNode} root
# @return {Integer[]}
def find_mode(root)
  initial_state = {row: 0, max_row: 0, current: nil, winners: []}
  dfs(root).each_with_object(initial_state) do |node, state|
    number = node.val
    if number == state[:current]
      state[:row] += 1
    else
      state[:row] = 1
      state[:current] = number
    end
    state[:winners] << number if state[:row] == state[:max_row]
    if state[:row] > state[:max_row]
      state[:max_row] = state[:row]
      state[:winners] = [number]
    end
  end[:winners]
end

File: elixir/find-mode-in-binary-search-tree.ex
# Definition for a binary tree node.
#
# defmodule TreeNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           left: TreeNode.t() | nil,
#           right: TreeNode.t() | nil
#         }
#   defstruct val: 0, left: nil, right: nil
# end

defmodule Solution do
  @spec find_mode(root :: TreeNode.t | nil) :: [integer]
  def find_mode(root),
    do:
      count(root)
      |> IO.inspect
      |> Enum.reduce(
        {[], 0},
        fn
          {val, count}, {_, mode_count} when count > mode_count -> {[val], count}
          {_, count}, {_, mode_count} = acc when count < mode_count -> acc
          {val, count}, {modes, mode_count} when count == mode_count -> {[val | modes], count}
        end
      )
      |> elem(0)

  def count(nil), do: %{}
  def count(%{val: val, left: left, right: right}),
    do:
      Map.merge(count(left), count(right), fn _k, v1, v2 -> v1 + v2 end)
      |> Map.update(val, 1, &(&1 + 1))
end
